runtime/near-vm-logic/src/dependencies.rs

```diff
@@ -48,6 +48,42 @@ pub trait ValuePtr {
 
 /// An external blockchain interface for the Runtime logic
 pub trait External {
+    /// Computes pairing check for alt_bn128 curve
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - serialized pairs of (sign, exp, base) points in LE notation
+    ///
+    /// # Example
+    /// ```
+    /// # use near_vm_logic::mocks::mock_external::MockedExternal;
+    /// # use near_vm_logic::External;
+    ///
+    /// # let mut external = MockedExternal::new();
+    /// let result = external.alt_bn128_g1_multiexp(...).unwrap(); // TODO fill example values
+    /// assert_eq!(&result, ...); // TODO fill example values
+    ///
+    /// ```
+    fn alt_bn128_g1_multiexp(&self, data: &[u8]) -> Result<Vec<u8>>;
```

Comment:
Anyway, we should check math objects passed to the VM (scalars should be in fields, points should be on curves).
Do we really need to start using complex types here (taking into account, that validation complexity is more than deserialization)?
Also, if we use serialized data to describe VM logic, in the future we could replace parity-bn to something else without the fork.

Reply:
> Do we really need to start using complex types here (taking into account, that validation complexity is more than deserialization)?

We need to deserialize the blob of bytes into a strongly typed objects in some place. We prefer to use strongly typed objects as early as possible, in this case it would been deserializing blobs right next to the Wasm bindings, which is in VMLogic. Not sure how the fact that we need to do complex validation supports or disproves the argument of where it should be done.

> Also, if we use serialized data to describe VM logic, in the future we could replace parity-bn to something else without the fork.

I don't think it is relevant to forking. The choice of serializing data in VMLogic on in the implementers of the External trait is pure architectural and has no effect on the protocol or the contracts written on our blockchain.

Reply:
Rewritten to strongly typed.